src/lib/Builder.hs

```diff
@@ -708,15 +708,16 @@ buildCase :: (Emits n, ScopableBuilder m)
           -> (forall l. (Emits l, DExt n l) => Int -> [AtomName l] -> m l (Atom l))
           -> m n (Atom n)
 buildCase scrut resultTy indexedAltBody = do
-  eff <- getAllowedEffects
   scrutTy <- getType scrut
   altsBinderTys <- caseAltsBinderTys scrutTy
-  alts <- forM (enumerate altsBinderTys) \(i, bs) -> do
-    buildNaryAbs bs \xs -> do
+  (alts, effs) <- unzip <$> forM (enumerate altsBinderTys) \(i, bs) -> do
+    blk <- buildNaryAbs bs \xs -> do
       buildBlock do
         ListE xs' <- sinkM $ ListE xs
         indexedAltBody i xs'
-  liftM Var $ emit $ Case scrut alts resultTy eff
+    eff <- effectsE blk
```

Comment:
The usage of `effectsE` in this patch follows a slightly annoying pattern where you often build up a bunch of `Alt`s and then have to open them up (through `refreshAbs`) in the implementation of `effectsE`. Now, since there are generally no emissions between their building and their analysis, this opening up should amount to checking that the names bound in the `Alt` are fresh, but it's still a check we have to do. It would be nice to consider if we could perhaps just call `effectsE` between `buildNaryAbs` and `buildBlock`.

Reply:
Thanks for the suggestion!  Moved extracting the block effect under `buildNaryAbs`.  No visible effect on allocations (perhaps none was expected), but seems to improve runtime (again, with the usual caveats about noise).

Not sure what you mean by "pattern", though.  This seems to be the only place in this patch where this happens?